File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.py
class Solution:
    def minSwaps(self, nums: List[int]) -> int:
        k = nums.count(1)
        if not k:
            return 0
        l, zero, ans, n = 0, 0, float("inf"), len(nums)
        nums *= 2

        for r in range(n * 2):
            if not nums[r]:
                zero += 1
            if r >= k - 1:
                ans = min(ans, zero)
                if not nums[l]:
                    zero -= 1
                l += 1
        return ans                     

File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.go
func minSwaps(nums []int) int {
    cnt := 0
    for _, v := range nums{
        if v == 1{
            cnt += 1
        }
    }
   

    nums = append(nums, nums...)
    left, cur := 0, 0
    ans := int(1e9)
    for right, val := range nums{
        if val == 0{
            cur += 1
        }
        if right - left + 1 > cnt {
            if nums[left] == 0{
                cur -= 1
            }
            left += 1
        }
        if right - left + 1 == cnt{
            ans = min(ans, cur)
        }     
    }
    return ans
}